src/gwt/src/org/rstudio/core/client/theme/DocTabLayoutPanel.java

```diff
@@ -923,42 +926,80 @@ public void endDrag(final Event evt, int action)
                   initDragParams_, null, destPos_));
          }
          
-         // this is the case when our own drag ends; if it ended outside our
-         // window and outside all satellites, treat it as a tab tear-off
-         if (dragElement_ != null && evt != null && action == ACTION_CANCEL)
+         if (Desktop.isDesktop())
          {
-            // if this is the last tab in satellite, we don't want to tear
-            // it out
-            boolean isLastSatelliteTab = getWidgetCount() == 1 && 
-                  Satellite.isCurrentWindowSatellite();
-
-            // did the user drag the tab outside this doc?
-            if (!isLastSatelliteTab &&
-                DomUtils.elementFromPoint(evt.getClientX(), 
-                  evt.getClientY()) == null)
+            // on desktop, we call back to discover whether the cursor is
+            // currently outside of any RStudio window; in such a case we
+            // perform a pop-out.
+            if (dragElement_ != null &&
+                  evt != null &&
+                  StringUtil.equals(evt.getType(), "dragend"))
             {
-               // did it end in any RStudio satellite window?
-               String targetWindowName;
-               Satellite satellite = RStudioGinjector.INSTANCE.getSatellite();
-               if (Satellite.isCurrentWindowSatellite())
-               {
-                  // this is a satellite, ask the main window 
-                  targetWindowName = satellite.getWindowAtPoint(
-                        evt.getScreenX(), evt.getScreenY());
-               }
-               else
-               {
-                  // this is the main window, query our own satellites
-                  targetWindowName = 
-                     RStudioGinjector.INSTANCE.getSatelliteManager()
-                         .getWindowAtPoint(evt.getScreenX(), evt.getScreenY());
-               }
-               if (targetWindowName == null)
+               Desktop.getFrame().doesWindowExistAtCursorPosition(
+                     new CommandWithArg<Boolean>()
+                     {
+                        @Override
+                        public void execute(final Boolean hasWindow)
+                        {
+                           if (hasWindow)
+                              return;
+
+                           Desktop.getFrame().getCursorPosition(new CommandWithArg<Point>()
```

Comment:
Nit: More succinct as a lambda. I think IntelliJ is starting to emit warnings for this pattern.